api/lib/opentelemetry/trace/span.rb

```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+
+# Copyright 2019 OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Trace
+    # Span is a mutable object storing information about the current operation execution.
+    class Span
+      attr_reader :context
+
+      def initialize(span_context: nil)
+        @context = span_context || SpanContext.new
+      end
+
+      def recording_events?
+        false
+      end
+
+      # TODO: API suggests set_attribute(key:, value:), but this feels more idiomatic?
+      def []=(key, value)
```

Comment:
I like the `[]=` approach over `set_attribute`, but I think it might be more clear if it were `attributes[]=` instead.

Reply:
I'm not so sure. An `#attributes` method could either return a `Hash`, which is too large an API—we don't want to imply that all `Hash` methods would be supported by the returned object, and force implementations either to use or duck-type the actual `Hash` class... or it could return a limited object that supports only `[]=`, but then we'd have to define that object as well, and that's also a headache for implementations. In the interest of a small API, I'm inclined to fall back on `set_attribute`, maybe with `[]=` as an alias.

Reply:
While I like the API `attributes[]=`, I agree that the implementation details to make that work, likely a slim wrapper around a hash (rather than an actual hash), are somewhat cumbersome for the implementor. Ultimately, I'd be fine with `set_attribute` and an alias of `[]=`.